editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/preferences/PreferredLabelEditor.java

```diff
@@ -0,0 +1,410 @@
+			item.setText(i, values[i]);
+		}
+		packColumns();
+	}
+
+	public void removeRows(int... indices) {
+		table.remove(indices);
+		packColumns();
+	}
+
+	protected void packColumns() {
+		for (TableColumn col : table.getColumns()) {
+			col.pack();
+		}
+	}
+
+	protected String[] getColumnHeaders() {
+		return new String[] { UIMessages.preferredLabelsComponentHeader, UIMessages.preferredLabelsParameterHeader };
+	}
+
+	/**
+	 * Returns this field editor's selection listener. The listener is created if nessessary.
+	 *
+	 * @return the selection listener
+	 */
+	private SelectionListener getSelectionListener() {
+		if (selectionListener == null) {
+			createSelectionListener();
+		}
+		return selectionListener;
+	}
+
+	/**
+	 * Returns this field editor's shell.
+	 * <p>
+	 * This method is internal to the framework; subclassers should not call this method.
+	 * </p>
+	 *
+	 * @return the shell
+	 */
+	protected Shell getShell() {
+		if (table == null) {
+			return null;
+		}
+		return table.getShell();
+	}
+
+	/**
+	 * Creates a selection listener.
+	 */
+	public void createSelectionListener() {
+		selectionListener = new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent event) {
+				Widget widget = event.widget;
+				if (widget == addButton) {
+					addPressed();
+				} else if (widget == editButton) {
+					editPressed();
+				} else if (widget == removeButton) {
+					removePressed();
+				} else if (widget == table) {
+					selectionChanged();
+				}
+			}
+		};
+	}
+
+	protected void addPressed() {
+		PreferredLabelDialog dialog = new PreferredLabelDialog(getShell());
+		dialog.setComponentValidator(new ComponentValidator(getComponents()));
+		dialog.setParameterValidator(new ParameterValidator());
+		if (dialog.open() == Window.OK) {
+			addRow(dialog.getComponent(), dialog.getParameter());
+		}
+	}
+
+	protected void editPressed() {
+		int index = table.getSelectionIndex();
+		if (index >= 0) {
+			TableItem selectedItem = table.getItem(index);
+			PreferredLabelDialog dialog = new PreferredLabelDialog(getShell());
+			dialog.setComponentValidator(new ComponentValidator());
+			dialog.setParameterValidator(new ParameterValidator());
+			dialog.setComponent(selectedItem.getText(0));
+			dialog.setParameter(selectedItem.getText(1));
+			if (dialog.open() == Window.OK) {
+				updateRow(index, dialog.getComponent(), dialog.getParameter());
+			}
+		}
+	}
+
+	protected void removePressed() {
+		removeRows(table.getSelectionIndices());
+	}
+
+	@Override
+	protected void doLoad() {
+		if (table != null) {
+			String s = getPreferenceStore().getString(getPreferenceName());
+			parseString(s);
+		}
+	}
+
+	@Override
+	protected void doLoadDefault() {
+		if (table != null) {
+			table.removeAll();
+			String s = getPreferenceStore().getDefaultString(getPreferenceName());
+			parseString(s);
+		}
+	}
+
+	@Override
+	protected void doStore() {
+		String s = createList(getPreferredLabels());
```

Comment:
what is "s"? The String representation of all the "preferredLabels"?
Can you provide a more meaningful name please.